packages/plugin-ext-vscode/src/node/plugin-vscode-resolver.ts

```diff
@@ -125,6 +143,40 @@ export class VsCodePluginDeployerResolver implements PluginDeployerResolver {
 
     }
 
+    /**
+     * Download and resolve the extension so that we can inspect the extensionDependencies
+     */
+    private async downloadVSCodeExtension(pluginResolverContext: PluginDeployerResolverContext, plugin: PluginDeployerEntry) {
+
+        const fileHandler = new PluginVsCodeFileHandler();
```

Comment:
hello, why a `new instance` here ? it's an injectable component as well as PluginVsCodeDirectoryHandler

I still don't understand why we're doing so much copy/paste stuff there over the plug-in deployment system, it looks like a big patch to avoid a clean implementation on taking care of dependencies

Reply:
The reason is because plugin-deployer-impl calls plugin-vscode-resolver so we cannot have plugin-vscode-resolver call plugin-deployer-impl to do the downloading etc otherwise there is a circular dependency.

Also, the injectable component is not available because multiple things are bound to PluginDeployerFileHandler. We could loop over everything bound to PluginDeployerFileHandler but I thought it would be better to just instantiate a new PluginVsCodeFileHandler because thats the only one relevant to it (This area of plugins I'm entirely new to)

Perhaps we can seperate the downloading of the plugin and the deployment of the plugin (basically just seperate plugin-deployer-impl into two different files) so that plugin-deploy-impl calls the downloading and so can the plugin-vscode-resolver? @benoitf WDYT?

Reply:
I think basic idea is to have classes doing minimal stuff at different stages so it's easy to plug new logic or to swipe to other logic.

Here, instantiating the classes directly without dependency injection is quite anti-pattern as some ppl may inject new stuff in the referenced classes but they should update as well clients so it's breaking dependency injection.

So yes, if the current model is not good enough, current classes need to be split again or new stages need to be added on top of the flow.

Abstraction is then: download, inspect, new dependencies ? then iterate again
then, sorting and deploy

Reply:
That makes sense, I'll work on making those changes! Thanks!

Reply:
@benoitf I was thinking that maybe we can have a queue of plugins to deploy here: https://github.com/theia-ide/theia/blob/741876b3ecba8a968ecf1200eb231639eb3b4074/packages/plugin-ext/src/main/node/plugin-deployer-impl.ts#L116

And we iterate till queue is not empty or only contains plugins which has been already deployed. After a plugin is processed by all components (resolver, file handler, folder handler) we take its dependencies and push them into the queue as well.

By dependencies i don't mean from package.json, but another attribute on plugin model in order to translate them into deployable form by one of components. For example the vscode folder handle can read dependencies from package.json and translate them into `vscode:extension/{extension-id}` dependencies on the plugin model.

What do you think?

@JPinkney please sync with @benoitf before implementing, it's his decision.

Reply:
@akosyakov yes I'm perfectly fine with pipeline queues

Reply:
BTW I'm more for consensus :-)

Reply:
@benoitf @akosyakov 
> And we iterate till queue is not empty or only contains plugins which has been already deployed. After a plugin is processed by all components (resolver, file handler, folder handler) we take its dependencies and push them into the queue as well.

Makes sense

> By dependencies i don't mean from package.json, but another attribute on plugin model in order to translate them into deployable form by one of components. For example the vscode folder handler can read dependencies from package.json and translate them into vscode:extension/{extension-id} dependencies on the plugin model.

Is this something that can be done when we are scanning the package.json? Can we add a method to the PluginScanner interface that converts an extensionDependency to a deployableExtensionDependency? and then every class that implements PluginScanner interface would implement the conversion that is neccessary. I.e. in scanner-vscode we would convert extensionDependencies to deployableExtensionDependencys by adding vscode:extension/ to every extensionDependency.

Reply:
@benoitf @akosyakov 
I went ahead with the modifications as best as I understood. extensionDependencies are now converted to deployableExtensionDependencies, the main plugin resolution now is a queue of plugins waiting to be installed with each plugin following resolve, applyFileHandlers, applyDirectoryHandler and then reading the deploy dependencies, any dependencies are then added into the queue